plugin: write Makefile recipes for `generate bundle` subcommand

estroz opened the pull request:
**Description of the change:**
Changes:
* `operator-sdk init` writes `config/bundle/kustomization.yaml` and adds `bundle` and `bundle-build` recipes that generate a bundle on disk and build a bundle image, respectively.
* `operator-sdk create api` manages `config/samples/kustomization.yaml` so `kustomize build config/bundle` retrieves custom resource samples.

**Motivation for the change:**

Example workflow:
```console
$ operator-sdk init
$ operator-sdk create api --group ship --version v1beta1 --kind Frigate
```
Then invoke:
```console
$ make bundle 
/home/estroz/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
kustomize build config/bundle | operator-sdk generate bundle -q --manifests --metadata --overwrite --version 0.0.1
$ make bundle-build
docker build -f bundle.Dockerfile -t controller-manager-bundle:latest .
```

TODO:
- [x] comment everything.
- ~[x] better testdata for current and new project layouts.~
- ~[x] rewrite tests (currently broken) in BDD style using gomega/ginkgo.~
- [x] break out into several PR's: #3018, #3064, #3065

Reference PR: #2956 

/kind feature

joelanford commented:
Overall looks like a great start! A few comments:
- I'm not a _huge_ fan of introducing a new `update` subcommand for this. Is there a way that `operator-sdk update bundle` can be folded into `operator-sdk generate bundle`?

  I think one question we need to answer is this: Do we or users care about persisting the result of `operator-sdk update bundle -q`? If not, then can `operator-sdk generate bundle` do the marker parsing and CSV building in memory, so that only the base UI metadata needs to be persisted?

- IIRC, the bundle `manifests` directory has to have a file in it with a `clusterserviceversion.yaml` suffix. Do we need to split out `memcached-operator.bundle.yaml` into its constituent parts?

estroz commented:
@joelanford Perhaps it would be better to fold `update bundle` into `generate bundle`; when in-binary CSV generation is removed in favor of `kustomize build`, the UI metadata generation component in `generate bundle` can stick around.

As for CSV file suffix: do you mean that is a requirement in the SDK (`generate csv`) or elsewhere?

joelanford commented:
> @joelanford Perhaps it would be better to fold `update bundle` into `generate bundle`; when in-binary CSV generation is removed in favor of `kustomize build`, the UI metadata generation component in `generate bundle` can stick around.

I think the make target in this PR is pretty close to what I imagined. I'm just wondering if the entrypoint can be `operator-sdk bundle generate` and we use a flag or another argument to run the kustomize generation vs the bundle directory creation. For example:
- `operator-sdk bundle generate manifests`
- `operator-sdk bundle generate directory`

`kustomize build config/default | operator-sdk bundle generate manifests`:
- creates `./config/bundle/base/clusterserviceversion.yaml` if not present (this is the base containing only the non-derivable metadata). This is populated by prompting the user for input.
- creates `./config/bundle/patches/*` which are the CSV patches to add the CRD sections, the roles, deployment, alm-examples, etc. These are all derived from the Go APIs and from kustomize output from other `./config` kustomizations. We maybe could shell out to kustomize if nothing is present on STDIN?
- creates `./config/bundle/kustomization.yaml` (if not present) that ties together the above files and includes `./config/crd` so that the CRDs end up in the kustomize output for `config/bundle`.

`kustomize build config/bundle | operator-sdk bundle generate directory`:
- gets all of the bundle resources from STDIN (or maybe shells out to kustomize)
- deletes and then re-creates `<bundle-dir>/manifests/` with each of the bundle manifest YAMLs
- creates `<bundle-dir>/metadata/annotations.yaml` if not present (anything here we need to or can update automatically?)
- creates `<bundle-dir>/Dockerfile` if not present

So the make target could be:
```Makefile
bundle: manifests
	kustomize build config/default | operator-sdk generate bundle manifests
	kustomize build config/bundle  | operator-sdk generate bundle directory
```

> As for CSV file suffix: do you mean that is a requirement in the SDK (`generate csv`) or elsewhere?

I swear I read something about OLM or the operator-registry knowing what a bundle directory is by searching for a file with a `clusterserviceversion.yaml` suffix. But now I can't find that. 🤷‍♂️

estroz commented:
I think we're on the right track. I'd still like to initially separate the timing of metadata generation from actually creating the bundle, even if `generate bundle` has the ability to do both, while allowing an interactive prompt at later stages in case a kubebuilder user is migrating to operator-sdk. More importantly though I want to provide both a simple and deeper OLM/bundle workflow. I've modified an example workflow from my [comment](https://github.com/operator-framework/enhancements/pull/16#discussion_r410497832) to illustrate a simplified bundle experience:

### Normal kubebuilder workflow

The following few commands are no different than the normal kubebuilder workflow; their behavior differs slightly:

Interactive prompt for UI metadata on `init`:
```console
$ operator-sdk init
Interactive CSV metadata prompt. Please fill in the following UI fields:
DisplayName: Memcached Operator
...
Icon file path: path/to/icon.png
```

Create APIs:
```console
$ operator-sdk create api --group cache --version v1 --kind Memcached
```

Create or update all CRD bases after Go types are created or modified, and update the CSV base:
```console
$ make manifests
controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle --update-bases --interactive=false
```

### Additional bundle workflow

Here's where we add to the normal kubebuilder workflow, after we've added APIs. We can run the following SDK-specific `make` recipes to work with bundles. Each recipe except `run-olm` gives an interactive prompt if a bundle kustomize base does not exist:

Generate a bundle on disk:
```sh
$ make manifests-olm
## Calls the 'manifests' recipe
controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle --update-bases
## Generate the bundle and metadata.
kustomize build config/olm-catalog | operator-sdk generate bundle --output-dir config/olm-catalog
```

Create a bundle image:
```sh
$ make bundle-build IMG=quay.io/example/memcached-operator:v0.0.1
## Calls the 'manifests' recipe
controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle --update-bases
## Creates a bundle in-memory then builds an image.
kustomize build config/olm-catalog | operator-sdk generate bundle --stdout | operator-sdk bundle create $(IMG)
```

Run directly with OLM:
```sh
$ make run-olm
## Calls the 'manifests' recipe
controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle --update-bases --interactive=false
## Creates a bundle in-memory then installs it with OLM.
kustomize build config/olm-catalog | operator-sdk generate bundle --stdout | operator-sdk run --olm
```

The reason `generate bundle` and `bundle create` still exist in my example is to allow fine-grained control over each step of the process if you want it. If you want things to "just work", then use any of those three `make` recipes. Once we move to `kustomize build` to generate the entire CSV from a base `generate bundle --stdout/--output-dir` will no longer be needed and that functionality can be removed; `generate bundle --update-bases` will still exist as the default behavior.

One drawback is a potentially confusing CLI if you run `operator-sdk --help`. I think we can mitigate confusion by adding a message like `Most users will only need 'make <recipe>'. Use this command directly only if you know what you are doing`.

estroz commented:
/cc @dmesser @shawn-hurley 

Thoughts on the above two CLI proposals?

dmesser commented:
Thanks for the example in context @estroz . I wonder if there is potential to avoid having two flavors with `operator-sdk generate bundle (manifests|directory)`. It will be hard to understand where the difference is and in what order they should be run. Ideally we could have one `generate bundle` command that does the kustomize base/patch and CSV generation and creating the bundle on disk structure. In both examples you show these commands run in consecutive order, so unless there is a use scenario where something needs to happen inbetween, we could save interaction surface and just provide one command.

Also I would favor `make manifests` always creating OLM bundles as well. Being part of the Operator Framework we would want to have integration by default and provide opt-out, not the other way around. if somebody doesn't want to use the bundle created by `make manifest` they do not have to.

shawn-hurley commented:
> Also I would favor make manifests always creating OLM bundles as well.
+1 

>  It will be hard to understand where the difference is and in what order they should be run. Ideally we could have one generate bundle command that does the kustomize base/patch and CSV generation and creating the bundle on disk structure.

+1 But I don't want us wrapping kustomize in the SDK. If we need to do two commands can we make the commands named more clearly what they are doing, for instance, `base-gen` instead of `generate bundle --update-bases`. WDYT?

dmesser commented:
> +1 But I don't want us wrapping kustomize in the SDK. If we need to do two commands can we make the commands named more clearly what they are doing, for instance, `base-gen` instead of `generate bundle --update-bases`. WDYT?

Why would we avoid wrapping/vendoring in kustomize in the SDK? It's an implementation we choose and there is no standalone use case for the rendered kustomize output.

joelanford commented:
> Why would we avoid wrapping/vendoring in kustomize in the SDK? It's an implementation we choose and there is no standalone use case for the rendered kustomize output.

Are we sure that's the case? What if I want to add some extra kustomize patches to my CSV that are not generated by default or that are specific to my organization/operator?

Why did Kubebuilder choose kustomize instead of rendering the final output directly?

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd.go:
```diff
@@ -39,10 +39,23 @@ are used to manage operators in the Operator Lifecycle Manager.
 	}
+	return cmd
+}
 
```
```suggestion
// todo: remove when the legacy layout is no longer supported
// Deprecated: Used just for the legacy layout
// --
// NewLegacyCmd will ...
```

camilamacedo86 commented on cmd/operator-sdk/bundle/generate.go:
```diff
@@ -0,0 +1,322 @@
+	channels       string
+	defaultChannel string
+	overwrite      bool
+}
```
The above split let me think that may should have more structures. Example:

```go
type bundleGenerateCmd struct { //nolint:maligned
        ... 
        manifestOptions manifestOptions
        metadataOptions metadataOptions
        ...
}

... 
type metadataOptions struct {
         directory      string
	channels       string
	defaultChannel string
	overwrite      bool
}
```

estroz replied:
That would make sense if we had methods only using `metadataOptions` fields, but all methods using `channels`, `defaultChannel`, and `overwrite` also use other fields from `bundleCmd`.

camilamacedo86 commented on cmd/operator-sdk/cli/legacy.go:
```diff
@@ -73,7 +73,7 @@ func GetCLIRoot() *cobra.Command {
 		alpha.NewCmd(),
 		build.NewCmd(),
-		bundle.NewCmd(),
+		bundle.NewLegacyCmd(),
```
May another option would be we keep the NewCmd and then do something as:

```
          if kbutil.HasProjectFile() {
		//todo actions
		return nil
	}

	// todo: remove when the legacy layout is no longer supported
	// note that the above if will no longer be required as well.
	if err := doLegacyX(..); err != nil {
		log.Fatalf("Failed to X %s: %v", n, err)
	}
	return nil
```

estroz replied:
You're suggesting using `kbutil.HasProjectFile()` in each `NewCmd()` that has both legacy and new behavior?

camilamacedo86 commented on internal/plugins/golang/init.go:
```diff
@@ -51,3 +59,83 @@ func (p *initPlugin) Run() error {
+	return ioutil.WriteFile("Makefile", makefileBytes, 0644)
+}
+
+func replaceOrAppendMakefileRecipe(oldMakefile []byte, recipeName, fragment string) (newMakefile []byte) {
```
You are customizing here the Makefile gen by the plugin. Am I right? 
So, it may be generic (customizeBullerplate) and added to kbutils since could be reused by others until we have the second phase. WDYT?

estroz replied:
Yes that's what's happening here. Probably a good idea to generalize this. I'm looking around for a makefile parser first because this update function isn't the most robust.

camilamacedo86 replied:
I do not think that we should update the current targets. I mean, we can just add at the end of the file the new bundle targets.

estroz commented:
Going to update this with code from #2956.

estroz commented on internal/generate/clusterserviceversion/bases/markers.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/operator-framework/operator-sdk/internal/generate/olm-catalog/descriptor"
+)
+
+func updateDescriptionsForGVKs(csv *operatorsv1alpha1.ClusterServiceVersion, apisDir string,
```
This was copied verbatim from [here](https://github.com/operator-framework/operator-sdk/blob/v0.17.0/internal/generate/olm-catalog/csv_updaters.go#L289).

estroz commented on internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+// apply applies the manifests in the collection to csv.
+func apply(c *collector.Manifests, csv *operatorsv1alpha1.ClusterServiceVersion) error {
```
Most of this file was copied verbatim from [here](https://github.com/operator-framework/operator-sdk/blob/v0.17.0/internal/generate/olm-catalog/csv_updaters.go)

estroz commented:
/cc @joelanford @dmesser @shawn-hurley

camilamacedo86 commented on cmd/operator-sdk/cli/legacy.go:
```diff
@@ -73,11 +73,11 @@ func GetCLIRoot() *cobra.Command {
 		alpha.NewCmd(),
 		build.NewCmd(),
-		bundle.NewCmd(),
+		bundle.NewCmdLegacy(),
```
May would be better to keep `bundle.NewCmd(),` and inside of it we identify if is the new layout or the old one as it has been done for the other cases.

estroz replied:
Well we've already centralized the check for layout in `main.go`, which will determine which commands are used. We shouldn't re-check layout in subcommands.

camilamacedo86 replied:
I see now. (`cmd/operator-sdk/cli/legacy.go` x `cmd/operator-sdk/cli/cli.go`)

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+	overwrite      bool
+}
+
+//nolint:lll
```
Why `//nolint:lll`? This lint means that the line is too extended. Is it really required in this scenario?

estroz replied:
I'll be adding some links in the long description that will be over 118 characters :grimacing:.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+			if !fs.Changed("kustomize") && !fs.Changed("metadata") && !fs.Changed("manifests") {
+				c.kustomize = true
+				c.manifests = true
+				c.metadata = true
```
If all is by default true so, why not set the default value equals true when the flag is defined [here](https://github.com/operator-framework/operator-sdk/pull/2860/files#diff-0aff0dcd8bfc140fd01b40d46a1f7bc1R99)?

estroz replied:
If no `kustomize`, `metadata`, or `manifests` flag is set, this command does nothing. The command is intended to generate a bundle, a process involving all 3 of these steps. Therefore, if the user sets no flag, I take that to mean "do the thing this command is supposed to". Turning off functionality with a boolean flag can be annoying if there are too many booleans:

If flag defaults are set to true:

```sh
# To only generate metadata
$ operator-sdk generate bundle --kustomize=false --manifests=false
# To generate kustomize and metadata
$ operator-sdk generate bundle --manifests=false
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle
```

If flag defaults are set to false:

```sh
# To only generate metadata
$ operator-sdk generate bundle --metadata
# To generate kustomize and metadata
$ operator-sdk generate bundle --metadata --kustomize
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle --metadata --kustomize --manifests
```

If options are set to true in `RunE` (current):

```sh
# To only generate metadata
$ operator-sdk generate bundle --metadata
# To generate kustomize and metadata
$ operator-sdk generate bundle --kustomize --metadata
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle
```

I prefer the latter, wdyt?

dmesser replied:
I prefer the latter, too/

camilamacedo86 replied:
Actually looking your explanation we do not really care about the bool values. Am I right? I mean, if it has the flag --metadata than should gen metadata, if has the --kustomize --metadata should gen both and we follow in this way.

So, could not it be simplified by checking if the flag was or not used instead of to set true or false in the middle of the scenarios as done above? if you see any reasons for not, wdyt about at least add a comment on top of this if and in any other place that you set true/false to manipulate it in order to described why/purpose.

estroz replied:
If I'm understanding you correctly, you can still set `--kustomize=false`, so only checking usage isn't enough. Can you provide an example of what you mean?

camilamacedo86 replied:
Just to clarifies. Note that, I am as a user I'd expected:

- To have a "bundle" with ALL requirements: > `$ operator-sdk generate bundle`
- To disable some option to gen this "bundle": `$ operator-sdk generate bundle --skip-option` OR `$ operator-sdk generate bundle --option=false`.  

Currently, we have, for example:  

> cmd.Flags().BoolVar(&c.metadata, "metadata", false, "Generate bundle metadata and Dockerfile")

According to the above explanation means: 

```
> # To only generate the metadata
> $ operator-sdk generate bundle --metadata
```

The above command would lead to me think that it will do **a full bundle +/with a metadata option**. However, the command will **ONLY** exec this specific piece instead of it. 

So, it is all fine if the current behaviour make sense for you folks . And then, in this situation, my suggestion would just improve the code comment and flags description to make its behaviour clear enough. E.g : 

```go
// Turn them all on by default here to simplify individual flag usage.
// # To only generate metadata
// $ operator-sdk generate bundle --metadata
// # To generate kustomize and metadata
// $ operator-sdk generate bundle --kustomize --metadata
// # To generate kustomize, manifests, metadata
// $ operator-sdk generate bundle
fs := cmd.Flags()
if !fs.Changed("kustomize") && !fs.Changed("metadata") && !fs.Changed("manifests") {
```
AND

```go
 // PS.: we need to clarifies it. Otherwise, is not clear why not just run `make kustomize instead of it`
cmd.Flags().BoolVar(&c.kustomize, "kustomize", false, "If set, it will generate/update ONLY the kustomize config bases with the specific bundle manifests")
cmd.Flags().BoolVar(&c.manifests, "manifests", false, "If set, it will generate/update ONLY the bundle manifests directory")
cmd.Flags().BoolVar(&c.metadata, "metadata", false, "If set, it will generate/update the bundle metadata directory and its Dockerfile")
```

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+				return fmt.Errorf("error reading configuration: %v", err)
+			}
+			c.setDefaults(cfg)
+
```
validate the config and manifest are all validations for the func so wdyt about 

```
cmd := &cobra.Command{
	if err = c.validate(cfg); err != nil {
		return fmt.Errorf("unable to run the comman: %v", err)
	}

        // todo exec command 
 
}
```
And then, have the validate with both. Also, any future impl to validated will be on it.

camilamacedo86 commented on internal/plugins/golang/init.go:
```diff
@@ -51,3 +59,86 @@ func (p *initPlugin) Run() error {
+bundle-build: manifests
+	kustomize build config/bundle | operator-sdk generate bundle -q --manifests --metadata --overwrite --version $(VERSION)
+	operator-sdk bundle validate config/bundle
+	docker build -f bundle.Dockerfile -t $(BUNDLE_IMG) .
```
Why we are changing the make file?
I do not think that we should impl the commands on it since we need to keep the operator-sdk commands ones since the make is not supported in win.

estroz replied:
To align with how kubebuilder implements functionality outside the core scaffold commands. Instead of having to run a bunch of intermediate steps to build a bundle image, a recipe like `bundle-build` captures all commands required to build that image in one `make` invocation.

camilamacedo86 replied:
However, will not we have the sdk cli command for that? So, why change the makefile? 
Shows that the make do not work for win SO, so shows that we should not really relay on that.

estroz replied:
The SDK's CLI is changing to accommodate `kustomize` usage, implying that most manifests in `config/` are not the "end" manifest like those the SDK scaffolds in `deploy/`. We could wrap `kustomize` in `operator-sdk`, but that isn't the right call order in my opinion. We should be sticking with kubebuilder's `make` flow.

The makefile doesn't work for Windows?

joelanford replied:
There was some debate with @shawn-hurley about keeping things like `operator-sdk build` around to accommodate Windows users who don't have `make`. But it does look like make can be installed natively in Windows (see [here](https://superuser.com/questions/808807/using-make-from-windows-powershell)), so maybe that fear was unfounded.

In any case, I second the Makefile approach. Our CLI has been too opinionated in some cases, and I think this approach (simple ready-made recipes in the Makefile + simple, do-one-thing SDK subcommands) strikes a nice balance.

camilamacedo86 replied:
Agreed.

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd.go:
```diff
@@ -44,10 +44,22 @@ Operator Lifecycle Manager:
+	return cmd
+}
 
+func NewCmdLegacy() *cobra.Command {
```
Could we add a deprecated note/comment?

estroz replied:
This isn't deprecated yet, and the linter will complain if an exported function is used while deprecated. Ditto below.

camilamacedo86 replied:
Regards the lint we can use nolint in this scenario, however, the exported one I agree that we should not add the note to not be visible in the godocs. 

However, I understand that we also agreed in the -2 E.g meeting ago use the Go deprecation annotation to make simple and very visible since it shows with strikethrough on the func and let us know very easy what should or not receiving a new logic and is used injust for the legacy and etc in order to eep the code maintained and cleanup all when the legacy one is no longer be supported.  

IMO the deprecated comment would be more important to be used in the funcs which are just used by `NewCmdLegacy()` and has no utility for the new layout at all.

camilamacedo86 replied:
Regards the lint we can use nolint in this scenario, however,  I agree that we should not add the note to not be visible in the godocs in the exported ones. 👍

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd.go:
```diff
@@ -44,10 +44,22 @@ Operator Lifecycle Manager:
+func NewCmdLegacy() *cobra.Command {
+	cmd := newCmd()
 	cmd.AddCommand(
 		newCreateCmd(),
```
the newCreateCmd is just used for the legacy one. Could we add a deprecated note on that as well?

camilamacedo86 replied:
So, IMO the deprecated comment would be more important to be used in the funcs which are just used by `NewCmdLegacy()` and has no utility for the new layout at all. If it is just used for the legacy one then,  in POV would make sense we deprecated it. 

PS.:  I understand that we agreed in the -2 ago eng. meeting use the Go deprecation annotations for all that is just used for the legacy one unless it make visible to the users since you are right we did not officially deprecate them. The pros here are let us know very easy what should or not receiving a new logic, what is or not just used for the legacy and etc in order to keep the code maintained and to help us clean up all when the legacy one is no longer be supported. Note that it makes very visible because shows a strikethrough on the func in the IDEs as well.

camilamacedo86 replied:
So, IMO the deprecated comment would be more important to be used in the funcs which are just used by `NewCmdLegacy()` and has no utility for the new layout at all. If it is just used for the legacy one then,  in POV would make sense we deprecated it.

hasbro17 commented:
@estroz Can you clarify the workflow for regenerating or updating the bundle manifests for an existing CSV version?

I can generate the bundle manifests the fist time:
```console
$ operator-sdk create api --group ship --version v1beta1 --kind Frigate
...
$ make bundle
/Users/haseeb/work/go-space/bin//controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
kustomize build config/bundle | operator-sdk generate bundle -q --manifests --version 0.0.1
```

At this point I change my sample CR `config/samples/ship_v1beta1_frigate.yaml` and want to update my CSV's `alm-example` field with the new sample.

Running `make bundle` again gives me the following validation error:
```console
$ make bundle
/Users/haseeb/work/go-space/bin//controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
kustomize build config/bundle | operator-sdk generate bundle -q --manifests --version 0.0.1
ERRO[0005] ClusterServiceVersion validation: [CSVFileNotValid] (app-operator.vX.Y.Z) metadata.name "app-operator.vX.Y.Z" is invalid:
[a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]
FATA[0005] Error generating bundle manifests: error generating ClusterServiceVersion: got ClusterServiceVersion validation errors
make: *** [Makefile:108: bundle] Error 1
```

Am I supposed to specify the CSV version in my `config/bundle/bases/app-operator.clusterserviceversion.yaml` base file?
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  annotations:
    alm-examples: '[]'
    capabilities: Basic Install
  name: app-operator.vX.Y.Z
  namespace: placeholder
```
I thought that's what the `--version 0.0.1` flag was for. Or is `make bundle` only a one time target to scaffold the manifests directory?

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+	// Metadata options.
+	channels       string
+	defaultChannel string
+	overwrite      bool
```
The above splits lead to me think that each one could be an structure.

estroz commented:
@hasbro17 that error is a regression following a recent change, will fix. `make bundle` should be idempotent.

@camilamacedo86 I can probably split this up further, will investigate.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+	return nil
+}
+
+func (c bundleCmd) runManifests(cfg *config.Config) (err error) {
```
We have been telling for the users use the `make` commands in all docs.
So, why should we port the make manifest and etc for go? Why not just call here the make commands instead of it?

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+	return nil
+}
+
+func (c bundleCmd) validateManifests(*config.Config) (err error) {
```
Same here. Why not just call `make manifest`?

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	if !c.quiet {
+		fmt.Println("Bases generated successfully")
```
```suggestion
		fmt.Println("Bundle manifest kustomize bases generated  successfully in: %v", c.outputDir)
```
It is important to know where the files have been added. 
Also, the nomenclature adopted here is not the same over above.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+		return fmt.Errorf("error generating ClusterServiceVersion: %v", err)
+	}
+
+	if err := kustomize.WriteIfNotExist(c.outputDir, kubebuilderKustomization); err != nil {
```
Are we not here addding bundle metafest files to allow customizations?
```suggestion
	if err := kustomize.WriteIfNotExist(c.outputDir, bundleKustomization); err != nil {
```

estroz replied:
No, `make bundle` runs `kustomize build` on those directories so any changes (ex. patches) a user wants applied to their manifests occur before being read by `generate bundle`.

camilamacedo86 replied:
May is valid check if the code can be more intuitive otherwise add comments to clarifies. Note that, my concern here is more about manutence ability. I mean,  be sure that any person that gets this code is able to understand what/where/why. e.g address in the code common/possible questions that will make for who tries to do a fix or rfe in this area: what diff has the kustomize implemented here with the default make one? Are we creating new files specifics for the bundle or just using the preexistent ones?

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+	}
+}
+
+func (c bundleCmd) runKustomize(cfg *config.Config) error {
```
`runKustomize ` -> led to me think that it will run a command such as : `kustomize build config/crd | kubectl apply -f -` however, is not this method generating all files indeed the CSV to build the bundle?

estroz replied:
This command effectively runs `generate bundle --kustomize`, hence the name. I will add a comment to clarify.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	if !c.quiet && !c.stdout {
+		fmt.Println("Bundle manifests generated successfully")
```
it is very helpful to know what and where was the file generated. Could we improve the log info?

hasbro17 replied:
@estroz and @camilamacedo86 We should consolidate further review comments on #3065 so we don't end up suggesting the same changes in both places.

@camilamacedo86 This is from your earlier suggestion to break this PR into smaller PRs.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -0,0 +1,217 @@
+- ../default
+- ../samples
+`
+
```
missing doc/comment for all funcs implemented.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+		},
+	}
+
+	cmd.Flags().BoolVar(&c.kustomize, "kustomize", false, "Generate kustomize bases")
+	cmd.Flags().BoolVar(&c.manifests, "manifests", false, "Generate bundle manifests")
+	cmd.Flags().BoolVar(&c.metadata, "metadata", false, "Generate bundle metadata and Dockerfile")
+	cmd.Flags().BoolVar(&c.stdout, "stdout", false, "Write bundle manifest to stdout")
```
// PS.: we need to clarifies it. Otherwise, is not clear why not just run `make kustomize instead of it`
```suggestion
	cmd.Flags().BoolVar(&c.kustomize, "kustomize", false, "If set, it will generate/update ONLY the kustomize config bases with the specific bundle manifests")
	cmd.Flags().BoolVar(&c.manifests, "manifests", false, "If set, it will generate/update ONLY the bundle manifests directory")
	cmd.Flags().BoolVar(&c.metadata, "metadata", false, "If set, it will generate/update the bundle metadata directory and its Dockerfile")
```

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+
+// TODO(estroz): add flag to skip API metadata regeneration.
+func (c *bundleCmd) addCommonFlagsTo(fs *pflag.FlagSet) {
+	fs.StringVar(&c.operatorName, "operator-name", "", "Name of the bundle's operator")
```
Would not it be the name of the operator IN the bundle?

estroz replied:
What is written and your suggestion mean the same thing. If that isn't clear I can update.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/cmd.go:
```diff
@@ -0,0 +1,123 @@
+	fs.StringVar(&c.operatorName, "operator-name", "", "Name of the bundle's operator")
+	fs.StringVarP(&c.version, "version", "v", "", "Semantic version of the operator in the generated bundle. "+
+		"Only set if creating a new bundle or upgrading your operator")
+	fs.StringVar(&c.inputDir, "input-dir", "", "Directory to read an existing bundle format from")
```
Could we add the info that is used when not set by default? IHMO It also clarifies what is expected to inform when required. 

```suggestion
	fs.StringVar(&c.inputDir, "input-dir", "", "Directory to read an existing bundle format from. If not set, it will be the path <N> by default.  )
```

estroz replied:
Setting defaults in flag help can get messy because you have to perform defaulting logic in the command constructor itself. If we decide to do so, this should be in a followup.

camilamacedo86 replied:
Ah ok, so it is not getting the values by default from the project scaffolded layout. Can we just improve the info to clarifies by giving examples over what is expected? Eg. `Directory to read an existing bundle format from. Eg <default path of it in the scaffolded project>`

camilamacedo86 commented on cmd/operator-sdk/generate/cmd.go:
```diff
@@ -16,15 +16,29 @@ package generate
+	return cmd
+}
+
+func NewCmdLegacy() *cobra.Command {
```
```suggestion
// todo: remove when the legacy layout is no longer supported
--
// Deprecated: Used just for the legacy layout
// --
func NewCmdLegacy() *cobra.Command {
```

estroz replied:
Meta note: do not set `Deprecated` on exported functions still in use. Doing so will break the linter.

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion.go:
```diff
@@ -0,0 +1,240 @@
+		oldName = existing.GetName()
+	}
+
+	// If the new version is empty, either because a CSV is only being updated or
+	// a base was generated, no update is needed.
+	if newVer == "0.0.0" || newVer == "" {
+		return nil
+	}
+
+	// Set replaces by default.
+	// TODO: consider all possible CSV versioning schemes supported  by OLM.
+	if oldVer != "0.0.0" {
+		csv.Spec.Replaces = oldName
+	}
```
if it has not in the api lib, could we not make a comment here and track this need for we use the lib check instead of impl the specific OLM validations here?

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion.go:
```diff
@@ -0,0 +1,240 @@
+
+	// Set replaces by default.
+	// TODO: consider all possible CSV versioning schemes supported  by OLM.
+	if oldVer != "0.0.0" {
```
Could it not be solved by `if len(oldVer) > 0 and oldVer  != "0.0.0" {`? Would not it remove the todo?

estroz replied:
By versioning schemes I mean OLM dependency resolution schemes, ex. semver, semver-skippatch, and replaces.

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -0,0 +1,379 @@
+				Expect(g.Generate(cfg, opts...)).ToNot(HaveOccurred())
+				outputFile := filepath.Join(tmp, bundle.ManifestsDir, makeCSVFileName("memcached-operator"))
+				Expect(outputFile).To(BeAnExistingFile())
+				Expect(string(readFileHelper(outputFile))).To(MatchYAML(newCSVStr))
```
Missing multigroup test. It has an if in the gen code that is not covered.

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion_test.go:
```diff
@@ -0,0 +1,379 @@
+				opts := []Option{
+					WithWriter(&bytes.Buffer{}),
+				}
+				Expect(g.Generate(cfg, opts...)).To(MatchError(genutil.InternalError("getBase must be set")))
```
should we not check for it case if the files expected was or not gen as well?

estroz replied:
In a follow-up.

estroz commented:
This PR is ready for review

/cc @hasbro17 @joelanford @camilamacedo86

hasbro17 commented on internal/plugins/golang/api.go:
```diff
@@ -52,5 +63,21 @@ func (p *createAPIPlugin) Run() error {
+
+// makeCRFile returns a Custom Resource example file name in the same format
+// as kubebuilder's CreateAPI plugin for a gvk.
+func makeCRFile(gvk config.GVK) string {
```
Nit: Perhaps a more accurate name for this func. Or `crFileName()`
```suggestion
func getCRFileName(gvk config.GVK) string {
```

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:
```diff
@@ -212,7 +212,6 @@ func mutatingToWebhookDescription(webhook admissionregv1.MutatingWebhook) operat
 func applyCustomResources(c *collector.Manifests, csv *operatorsv1alpha1.ClusterServiceVersion) error {
 	examples := []json.RawMessage{}
 	for _, cr := range c.CustomResources {
-		fmt.Println("applying", cr.GetName())
```
Why the printed info was removed?

estroz replied:
I forgot to remove these from a previous PR.

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:
```diff
@@ -263,6 +261,8 @@ func validate(csv *operatorsv1alpha1.ClusterServiceVersion) error {
 	}
+
+	hasErrors := false
 	results := validation.ClusterServiceVersionValidator.Validate(csv)
```
```suggestion
	results := validation.ClusterServiceVersionValidator.Validate(csv)
	for _, r := range results {
		for _, w := range r.Warnings {
			log.Warnf("ClusterServiceVersion validation: [%s] %s", w.Type, w.Detail)
		}
		for _, e := range r.Errors {
		        hasErrors = true
			log.Errorf("ClusterServiceVersion validation: [%s] %s", e.Type, e.Detail)
		}
	}
```

camilamacedo86 commented on internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:
```diff
@@ -272,8 +272,13 @@ func validate(csv *operatorsv1alpha1.ClusterServiceVersion) error {
-			return errors.New("got ClusterServiceVersion validation errors")
+			hasErrors = true
 		}
 	}
```
```suggestion
	results := validation.ClusterServiceVersionValidator.Validate(csv)
	for _, r := range results {
		for _, w := range r.Warnings {
			log.Warnf("ClusterServiceVersion validation: [%s] %s", w.Type, w.Detail)
		}
		for _, e := range r.Errors {
		        hasErrors = true
			log.Errorf("ClusterServiceVersion validation: [%s] %s", e.Type, e.Detail)
		}
	}
```

camilamacedo86 replied:
has no need to `if r.HasError() {`

estroz replied:
This is doing effectively the same thing as checking `r.HasError()`, but with a slight performance loss because it's resetting memory for each error (assuming the compiler doesn't optimize this into one operation).

camilamacedo86 replied:
I  think here`slight performance loss` does not justify a would be more valuable looking for readability. However, 👍

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -69,6 +76,10 @@ func (c bundleCmd) runKustomize(cfg *config.Config) error {
+
 	if !c.quiet {
 		fmt.Println("Bases generated successfully in", c.outputDir)
 	}
```
It shows not be working. See: (I just follow up the quickstart steps and executed `make bundle`)

```
$ make bundle
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
kustomize build config/bundle | operator-sdk generate bundle -q --manifests --version 0.0.1
Error: accumulating resources: couldn't make target for path '/Users/camilamacedo/go/src/test-make/config/samples': unable to find one of 'kustomization.yaml', 'kustomization.yml' or 'Kustomization' in directory '/Users/camilamacedo/go/src/test-make/config/samples'
```

<details>
<summary>
see here all steps`
</summary>

```
camilamacedo@Camilas-MacBook-Pro ~/go/src $ cd test-make/
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-make $ operator-sdk init --domain my.domain
Writing scaffold for you to edit...
Get controller runtime:
$ go get sigs.k8s.io/controller-runtime@v0.6.0
Update go.mod:
$ go mod tidy
Running make:
$ make
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager main.go
Next: define a resource with:
$ operator-sdk create api
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-make $ kubebuilder create api --group webapp --version v1 --kind Guestbook
Create Resource [y/n]
y
Create Controller [y/n]
y
Writing scaffold for you to edit...
api/v1/guestbook_types.go
controllers/guestbook_controller.go
Running make:
$ make
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager main.go
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-make $ make install
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize

Display name for the operator (required): 
> test

Description for the operator (required): 
> test

Provider's name for the operator (required): 
> test

Any relevant URL for the provider name (optional): 
> 

Comma-separated list of keywords for your operator (required): 
> 
No list provided. 
Comma-separated list of keywords for your operator (required): 
> 
No list provided. 
Comma-separated list of keywords for your operator (required): 
> 
No list provided. 
Comma-separated list of keywords for your operator (required): 
> 
No list provided. 
Comma-separated list of keywords for your operator (required): 
> test,test

Comma-separated list of maintainers and their emails (e.g. 'name1:email1, name2:email2') (required): 
> name
,
/Users/camilamacedo/go/bin/kustomize build config/crd | kubectl apply -f -
customresourcedefinition.apiextensions.k8s.io/guestbooks.webapp.my.domain created
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-make $ ,
-bash: ,: command not found
camilamacedo@Camilas-MacBook-Pro ~/go/src/test-make $ make bundle
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
kustomize build config/bundle | operator-sdk generate bundle -q --manifests --version 0.0.1
Error: accumulating resources: couldn't make target for path '/Users/camilamacedo/go/src/test-make/config/samples': unable to find one of 'kustomization.yaml', 'kustomization.yml' or 'Kustomization' in directory '/Users/camilamacedo/go/src/test-make/config/samples'
```

camilamacedo86 commented on internal/plugins/golang/api.go:
```diff
@@ -52,5 +63,21 @@ func (p *createAPIPlugin) Run() error {
+	kpath := filepath.Join("config", "samples")
+	if err := kustomize.Write(kpath, samplesKustomization); err != nil {
+		return err
+	}
```
Should we really re-write the kb plugin? Would not be better we add a new our/plugin that will be executed after just to scaffold its requirement? something as ` (p *createBundlePlugin)`. If we follow up this approach can we not end up in a situation where we will change the kb plugins and then, both projects are no longer compatible?

estroz replied:
>add a new our/plugin that will be executed after just to scaffold its requirement

If I understand what you're asking, we are effectively doing that with:

```go
import (
	...
	"sigs.k8s.io/kubebuilder/pkg/plugin"
)

type createAPIPlugin struct {
	plugin.CreateAPI
	...
}

func (p *createAPIPlugin) Run() error {
	if err := p.CreateAPI.Run(); err != nil {
		return err
	}
	...
	return p.run()
}
```

This emulates phase 2 plugin behavior.

camilamacedo86 replied:
My idea here was that we would have a new plugin such as  `_ sdkplugin.BundlePluginGetter       = Plugin{}` and not overwrite the init one.  And then, the Bundle plugin when/if called, would update the Makefile scaffolds and etc which means that we could indeed just to do the customization if a command such as  `operator-sdk bundle create` be executed for example. 

**However, I am not sure if it will be easy to be done now .. so it is fine. Out of scope here.**  👍

estroz commented:
@camilamacedo86 you ran
```sh
$ kubebuilder create api --group webapp --version v1 --kind Guestbook
```
which will not scaffold a `config/samples/kustomization.yaml`. Only
```sh
$ operator-sdk create api --group webapp --version v1 --kind Guestbook
```
will do so. It would be better UX to scaffold this file when running `init` but `kustomize` does not let you glob all files in a directory in an out-of-tree referenced `kustomization.yaml`, so we have to overwrite it each time an API is created.

estroz commented:
@hasbro17 can you take another look at this? Changes:
* Added command help text and examples.
* `operator-sdk init` now scaffolds `config/bundle/kustomization.yaml` since you can have an operator without custom APIs

hasbro17 commented:
> init now scaffolds config/bundle/kustomization.yaml instead of generate bundle --kustomize

@estroz Don't you mean `"init" now scaffolds config/bundle/kustomization.yaml instead of "create api"`? 
From my understanding `generate bundle --kustomize` only generated `config/bundle/bases/...` and not `config/bundle/kustomization.yaml`.

camilamacedo86 commented:
Hi @estroz and @hasbro17, 

Following some points: 

**By following the QuickStart I should not care about the bundle** 

I am as a user would like to create the project and not a bundle. However, the basic steps are asking the bundle info when I run `make install`. See: 


<details>
<summary>
See here all steps
</summary>

```
camilamacedo@Camilas-MacBook-Pro ~/go/src $ mkdir pr-bundle
camilamacedo@Camilas-MacBook-Pro ~/go/src $ cd pr-bundle/
camilamacedo@Camilas-MacBook-Pro ~/go/src/pr-bundle $ operator-sdk init --domain my.domain
Writing scaffold for you to edit...
Get controller runtime:
$ go get sigs.k8s.io/controller-runtime@v0.6.0
Update go.mod:
$ go mod tidy
Running make:
$ make
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager main.go
Next: define a resource with:
$ operator-sdk create api
camilamacedo@Camilas-MacBook-Pro ~/go/src/pr-bundle $ operator-sdk create api --group webapp --version v1 --kind Guestbook
Create Resource [y/n]
y
Create Controller [y/n]
y
Writing scaffold for you to edit...
api/v1/guestbook_types.go
controllers/guestbook_controller.go
Running make:
$ make
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager main.go
camilamacedo@Camilas-MacBook-Pro ~/go/src/pr-bundle $ make install
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize

Display name for the operator (required): 
> 
```
</details>

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+
+  # Generate bundle files and build your bundle image with this 'make' recipe:
+  $ export USERNAME=<your registry username>
+  $ make bundle-build BUNDLE_IMG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
```
we have now 2 targets that will do it. IMO we should have only  `make bundle-build`. See: 

```
$  make bundle
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
/Users/camilamacedo/go/bin/kustomize build config/bundle | operator-sdk generate bundle -q --manifests --version 0.0.1
camilamacedo@Camilas-MacBook-Pro ~/go/src/pr-bundle $  make bundle-build
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize
/Users/camilamacedo/go/bin/kustomize build config/bundle | operator-sdk generate bundle -q --manifests --metadata --overwrite --version 0.0.1
INFO[0005] Building annotations.yaml                    
INFO[0005] Building Dockerfile                          
operator-sdk bundle validate config/bundle
INFO[0000] Found annotations file                        bundle-dir=config/bundle container-tool=docker
INFO[0000] Could not find optional dependencies file     bundle-dir=config/bundle container-tool=docker
INFO[0000] All validation tests have completed successfully 
docker build -f bundle.Dockerfile -t controller-bundle:0.0.1 .
Sending build context to Docker daemon  42.43MB
Step 1/9 : FROM scratch
 ---> 
Step 2/9 : LABEL operators.operatorframework.io.bundle.mediatype.v1=registry+v1
 ---> Using cache
 ---> 91093cb32610
Step 3/9 : LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
 ---> Using cache
 ---> 26e40a8a5669
Step 4/9 : LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
 ---> Using cache
 ---> c44955a2769a
Step 5/9 : LABEL operators.operatorframework.io.bundle.package.v1=pr-bundle
 ---> Using cache
 ---> aa8a3d023e0f
Step 6/9 : LABEL operators.operatorframework.io.bundle.channels.v1=alpha
 ---> Using cache
 ---> 519a771b893d
Step 7/9 : LABEL operators.operatorframework.io.bundle.channel.default.v1=alpha
 ---> Using cache
 ---> 396b25fd7c9f
Step 8/9 : COPY config/bundle/manifests /manifests/
 ---> Using cache
 ---> 65f1b9577b20
Step 9/9 : COPY config/bundle/metadata /metadata/
 ---> Using cache
 ---> 77c3fa71a9ee
Successfully built 77c3fa71a9ee
Successfully tagged controller-bundle:0.0.1
```

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+
+  # The above recipe runs the following commands manually. First it creates bundle
+  # manifests, metadata, and a bundle.Dockerfile:
+  $ make manifests
```
IMO we should not make mandatory for users to provide the bundle info to integrate the project with OLM in all cases

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+const (
+	//nolint:lll
+	examples = `
+  # Using the example 'memcached-operator' and assuming a directory structure
```
Is required to add the tree? What value does it bring?

estroz replied:
It isn't strictly necessary, although other commands show trees.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+  operator-sdk generate bundle -q --kustomize
+
+  Display name for the operator (required):
+  > memcached-operator
```
Is required to add its output?

estroz replied:
Nope, just helps the user know what expected output should be. I can see this drifting from actual output though.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+
+  # Then it validates your bundle files and builds your bundle image:
+  $ operator-sdk bundle validate config/bundle
+  $ docker build -f bundle.Dockerfile -t quay.io/$USERNAME/memcached-operator-bundle:v0.0.1 .
```
Should we not use here BUNDLE_IMG instead of `quay.io/$USERNAME/memcached-operator-bundle:v0.0.1`?

estroz replied:
We could. The example would then look like:
```sh
$ export BUNDLE_IMG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
$ docker build -f bundle.Dockerfile -t $BUNDLE_IMG
$ make docker-push BUNDLE_IMG=$BUNDLE_IMG
```

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+  $ operator-sdk bundle validate config/bundle
+  $ docker build -f bundle.Dockerfile -t quay.io/$USERNAME/memcached-operator-bundle:v0.0.1 .
+  Sending build context to Docker daemon  42.33MB
+  Step 1/9 : FROM scratch
```
Is required add the output?

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+  ...
+
+  # You can then push your bundle image:
+  $ make docker-push IMG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
```
```suggestion
  $ make docker-push BUNDLE_MG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
```

It should not use the IMG since will make it very confusing. See 

<img width="804" alt="Screenshot 2020-05-27 at 12 14 42" src="https://user-images.githubusercontent.com/7708031/83012698-ab7adc80-a013-11ea-8544-485dcb9e7451.png">

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+
+  # You can then push your bundle image:
+  $ make docker-push IMG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
+`
```
Can we use here BUNDLE_IMG as well? Just for not be confusing with the default instructions to build and push the operator images.

camilamacedo86 commented on internal/plugins/golang/init.go:
```diff
@@ -43,11 +50,118 @@ func (p *initPlugin) Run() error {
+	if err := kustomize.Write(filepath.Join("config", "bundle"), bundleKustomization); err != nil {
+		return err
+	}
+
```
Could we not centralize the code in a bundle.go as we have the config.go and here just call 1 method?

camilamacedo86 commented on internal/plugins/golang/init.go:
```diff
@@ -43,11 +50,118 @@ func (p *initPlugin) Run() error {
 }
+
+// kustomization for bundles.
+const bundleKustomization = `resources:
```
Could we add a note for we try to improve its structure and have something that will follow up the KB layout?

 E.g `pkg/scaffold/internal/templates/bundle/makefile.go` where would have the new bundle targets. Since we should not change the existent ones just add new we could add at the end the new targets in this scenario. 



However, it might increase the complexity. 



(Out of teh scope here, we can shape as required in the futire)

estroz replied:
These should live here now, then can be moved if we change anything else about the Makefile.

camilamacedo86 commented on cmd/operator-sdk/generate/bundle/bundle.go:
```diff
@@ -28,6 +28,76 @@ import (
+
+  # Generate bundle files and build your bundle image with this 'make' recipe:
+  $ export USERNAME=<your registry username>
+  $ make bundle-build BUNDLE_IMG=quay.io/$USERNAME/memcached-operator-bundle:v0.0.1
```
We have now 2 makefile target which make confuse. See:

```
# Build the bundle OCI image
bundle-build: manifests
	$(KUSTOMIZE) build config/bundle | operator-sdk generate bundle -q --manifests --metadata --overwrite --version $(VERSION)
	operator-sdk bundle validate config/bundle
	docker build -f bundle.Dockerfile -t $(BUNDLE_IMG) .

# Generate a bundle directory
bundle: manifests
	$(KUSTOMIZE) build config/bundle | operator-sdk generate bundle -q --manifests --version $(VERSION)
```

Could we not remove the make bundle and just keep the make build-bundle instead of it which has been used in the example? Why keep both?

estroz replied:
Having one to scaffold a bundle and another to build a bundle image seems straightforward no?

estroz commented:
@camilamacedo86 lets have UX discussions in https://github.com/operator-framework/enhancements/pull/16.

hasbro17 commented:
@estroz Actually I just tested what @camilamacedo86 pointed out and I'm not sure if this was always the case but after scaffolding a basic project without intending to write CSVs or bundles I'm still forced to input the metadata for the CSV bases on `make manifest` or `make install`:

```console
$ operator-sdk init --repo=github.com/hasbro17/memcached-operator --domain=example.com
...

$ operator-sdk create api --group cache --version v1alpha1 --kind Memcached
...

$ make manifests
/Users/haseeb/work/go-space/bin//controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize

Display name for the operator (required):
...
```

This kind of forces CSVs and bundles into the workflow by default rather than making it opt-in like we had with `generate csv` or `generate bundle` in the old layout.

hasbro17 commented:
@estroz Nvm just saw the comment for UX discussions. Ignore my previous comment.
